cryptolib/verifier.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cryptolib
+
+import "errors"
+
+// Verifier contains methods to validate an Attestation.
+type Verifier interface {
+	// VerifyAttestation verifies whether an Attestation satisfies at least one
+	// of the public keys under an image. This function finds the public key
+	// whose ID matches the attestation's PublicKeyID, and uses this key to
+	// verify the signature.
+	VerifyAttestation(att *Attestation, publicKeySet []PublicKey, image string) error
+}
+
+// PublicKey stores public key material for all key types.
+type PublicKey struct {
```

Comment:
How do I create a PublicKey?  Signer has NewThisSinger and NewThatSigner, which makes it nice and obvious how to create one, but there's no such obvious counterpart here.

Reply:
Done.

Reply:
OK, now I see what you're doing.  PublicKey doesn't "know" which keyMode it is.  Instead, when you verify an attestation, you deduce the keyMode from the attestation, and then try to parse the key.  This is different than the signer, where key mode is fixed at signer creation time (rather than later at signing time).

I wonder why the asymmetry?  It seems nicer to parse the key once, as early as possible.  It gives you some optimization opportunities, and also allows earlier error signaling.

Also -- supposing we ever add even more key types, will it always be possible to determine key type from attestation?

This is a non-blocking comment, if you want to stick to the current design, I don't feel strongly about it.  I just wonder if you'd thought about the alternative.